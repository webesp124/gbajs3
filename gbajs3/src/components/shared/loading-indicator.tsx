import { PacmanLoader } from 'react-spinners';
import { styled, useTheme } from 'styled-components';

import type { ReactNode } from 'react';
import type { LoaderSizeMarginProps } from 'react-spinners/helpers/props';

type LoadingIndicatorProps = {
  children: ReactNode;
  currentName: string | null;
  indicator: ReactNode;
  isLoading: boolean;
  loadingCopy: string;
};

const LoadingContainer = styled.div`
  display: flex;
  flex-direction: column;
  gap: 10px;
  text-align: center;
  align-items: center;
  margin-bottom: 15px;
`;

const Copy = styled.p`
  word-wrap: break-word;
  max-width: 100%;
`;

export const PacmanIndicator = (props: LoaderSizeMarginProps) => {
  const theme = useTheme();

  return (
    <PacmanLoader
      color={theme.gbaThemeBlue}
      cssOverride={{ margin: '0 auto' }}
      {...props}
    />
  );
};

export const LoadingIndicator = ({
  children,
  currentName,
  indicator,
  isLoading,
  loadingCopy
}: LoadingIndicatorProps) => {
  return isLoading ? (
    <LoadingContainer>
      <Copy>
        {loadingCopy}
        <br />
        {currentName}
      </Copy>
      {indicator}
    </LoadingContainer>
  ) : (
    children
  );
};
