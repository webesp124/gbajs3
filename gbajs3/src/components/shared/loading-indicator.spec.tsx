import { screen } from '@testing-library/react';
import { describe, it, expect } from 'vitest';

import { PacmanIndicator, LoadingIndicator } from './loading-indicator.tsx';
import { renderWithContext } from '../../../test/render-with-context.tsx';

describe('<PacmanIndicator />', () => {
  it('renders pacman indicator with additional props', () => {
    renderWithContext(
      <PacmanIndicator size={50} data-testid="pacman-indicator" />
    );

    expect(screen.getByTestId('pacman-indicator')).toMatchSnapshot();
  });
});

describe('<LoadingIndicator />', () => {
  it('renders indicator when isLoading is true', () => {
    renderWithContext(
      <LoadingIndicator
        isLoading
        loadingCopy="Loading..."
        currentName="Test"
        indicator={<PacmanIndicator data-testid="pacman-indicator" />}
      >
        <div>Child Content</div>
      </LoadingIndicator>
    );

    expect(screen.getByText(/Loading.../)).toBeInTheDocument();
    expect(screen.getByText(/Test/)).toBeInTheDocument();
    expect(screen.getByTestId('pacman-indicator')).toBeInTheDocument();
  });

  it('renders children when isLoading is false', () => {
    renderWithContext(
      <LoadingIndicator
        isLoading={false}
        loadingCopy="Loading..."
        currentName="Test"
        indicator={<PacmanIndicator data-testid="pacman-indicator" />}
      >
        <div>Child Content</div>
      </LoadingIndicator>
    );

    expect(screen.getByText('Child Content')).toBeInTheDocument();
    expect(screen.queryByText(/Loading.../)).not.toBeInTheDocument();
    expect(screen.queryByTestId('pacman-indicator')).not.toBeInTheDocument();
  });
});
