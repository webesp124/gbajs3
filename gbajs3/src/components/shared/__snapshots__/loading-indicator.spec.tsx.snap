// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<PacmanIndicator /> > renders pacman indicator with additional props 1`] = `
<span
  data-testid="pacman-indicator"
  style="display: inherit; position: relative; font-size: 0px; height: 100px; width: 100px; margin: 0px auto;"
>
  <span
    style="width: 0px; height: 0px; border-right: 50px solid transparent; border-top: 50px solid transparent; border-left: 50px solid #0d6efd; border-bottom: 50px solid #0d6efd; border-radius: 50px; position: absolute; animation: react-spinners-PacmanLoader-pacman-1 0.8s infinite ease-in-out; animation-fill-mode: both;"
  />
  <span
    style="width: 0px; height: 0px; border-right: 50px solid transparent; border-top: 50px solid #0d6efd; border-left: 50px solid #0d6efd; border-bottom: 50px solid transparent; border-radius: 50px; position: absolute; animation: react-spinners-PacmanLoader-pacman-2 0.8s infinite ease-in-out; animation-fill-mode: both;"
  />
  <span
    style="width: 16.666666666666668px; height: 16.666666666666668px; background-color: rgb(13, 110, 253); margin: 2px; border-radius: 100%; transform: translate(0, -12.5px); position: absolute; top: 50px; left: 200px; animation: react-spinners-PacmanLoader-ball 1s 0.5s infinite linear; animation-fill-mode: both;"
  />
  <span
    style="width: 16.666666666666668px; height: 16.666666666666668px; background-color: rgb(13, 110, 253); margin: 2px; border-radius: 100%; transform: translate(0, -12.5px); position: absolute; top: 50px; left: 200px; animation: react-spinners-PacmanLoader-ball 1s 0.75s infinite linear; animation-fill-mode: both;"
  />
  <span
    style="width: 16.666666666666668px; height: 16.666666666666668px; background-color: rgb(13, 110, 253); margin: 2px; border-radius: 100%; transform: translate(0, -12.5px); position: absolute; top: 50px; left: 200px; animation: react-spinners-PacmanLoader-ball 1s 1s infinite linear; animation-fill-mode: both;"
  />
  <span
    style="width: 16.666666666666668px; height: 16.666666666666668px; background-color: rgb(13, 110, 253); margin: 2px; border-radius: 100%; transform: translate(0, -12.5px); position: absolute; top: 50px; left: 200px; animation: react-spinners-PacmanLoader-ball 1s 1.25s infinite linear; animation-fill-mode: both;"
  />
</span>
`;
